[YANYAN-w] iP

YANYAN-w opened the pull request:
# Katheryne
> “Your mind is for having ideas, not holding them.” – David Allen

Katheryne is a powerful chatbot that can help you manage your daily tasks.

- text-based
- east to learn
- simple syntax
- ~fast~ **SUPER FAST** to use

To implement this powerful tool, all you need to do is:

1. download from [here](https://github.com/YANYAN-w/ip)
2. run it
3. add your tasks
4. let Kathryne manage your task 😻 

Features:

- [x] Manage tasks
- [ ] Manage deadlines 
- [ ] Reminder (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Katheryne {
    public static void main(String[] args) throws MissingInformationException {
        Katheryne k = new Katheryne("./data/Katheryne.txt");
            k.run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

YANYAN-w committed:
level 1-0

YANYAN-w committed:
Level-0

YANYAN-w committed:
Level-1

YANYAN-w committed:
Level-1

YANYAN-w committed:
Level-2

YANYAN-w committed:
Level-3

YANYAN-w committed:
no message

YANYAN-w committed:
add sub classes of Task

YANYAN-w committed:
edited response genereated and added test case with expected output.

YANYAN-w committed:
Handle exceptions.

YANYAN-w committed:
add delete function

YANYAN-w committed:
Add save() method

YANYAN-w committed:
Merge branch 'branch-Level-7'

YANYAN-w committed:
Edit string format for toSaveString() for deadline and event class;

YANYAN-w committed:
Level-8

YANYAN-w committed:
Merge branch 'branch-Level-8'

YANYAN-w committed:
Resturecture the OOP

YANYAN-w committed:
package

YANYAN-w committed:
Merge branch 'add-gradle-support'

YANYAN-w committed:
Add Junit test

YANYAN-w committed:
Add JavaDoc comments

YANYAN-w committed:
Update code to correct style

YANYAN-w committed:
Implement search function

YANYAN-w committed:
Update for output message

YANYAN-w committed:
Merge branch 'branch-Level-9'

YANYAN-w committed:
Merge branch 'branch-A-JavaDoc' to master

YANYAN-w committed:
Merge branch 'branch-A-CodingStandard'

YANYAN-w committed:
Katheryne class: updates in header comment

YANYAN-w committed:
Ui class: refine the getListMessage() method

YANYAN-w committed:
Command class: clean up code for stylecheck

YANYAN-w committed:
Message class: update repsonse for list function

YANYAN-w committed:
Update the print output for list function

YANYAN-w committed:
Add checkstyle plugin to detect coding style violations

YANYAN-w committed:
Merge branch 'branch-A-Checkstyle'

YANYAN-w committed:
Add a gui to the chatbot

YANYAN-w committed:
Merge branch 'branch-Level-10'

weibinlim commented on src/main/java/katheryne/Command.java:
```diff
@@ -16,10 +16,12 @@ public String executeList(TaskList l) {
     }
 
-    public String executeMark(String string) throws MissingInformationException, IndexOutOfBoundsException {
+    public String executeMark(String string)
+            throws MissingInformationException, IndexOutOfBoundsException {
```
I like how you placed a line break in the method signature to improve readability!

weibinlim commented on src/main/java/katheryne/Command.java:
```diff
@@ -42,53 +46,64 @@ public String executeUnmark(String string) throws MissingInformationException, I
         String[] fullCommand = Parser.parseCommand(string);
         if (fullCommand.length < 2 || fullCommand[1].isEmpty()) {
-            throw new MissingInformationException("You need to specify the description of an event.");
+            String msg = "You need to specify the description of an event.";
```
I like that your exception message is specific to the error!

weibinlim commented on src/main/java/katheryne/Deadline.java:
```diff
@@ -3,6 +3,12 @@
 import java.time.LocalDate;
 import java.time.format.DateTimeFormatter;
 
+/**
+ * Deadline Class is a child class of Task Class.
+ * It has an additional attribute of time, which specify
+ * the time of the deadline.
+ * The time attribute is a LocalDate object.
+ */
```
I like that you JavaDocs is very comprehensive!

weibinlim commented on src/main/java/Deadline.java:
```diff
@@ -1,21 +1,29 @@
+
 public class Deadline extends Task{
-    protected String by;
+    protected LocalDate time;
```
Perhaps you could use a different variable name to avoid confusion?

weibinlim commented on src/main/java/katheryne/Message.java:
```diff
@@ -16,4 +18,6 @@ public class Message {
     public static final String MESSAGE_DATE_FORMAT_ERROR = "Katheryne: Please enter date in 'yyyy-MM-dd' format";
     public static final String MESSAGE_INDEX_ERROR = "Katheryne: That task number doesn't exist. Please try again.";
     public static final String MESSAGE_LOADING_ERROR = "Katheryne: No saved task found. Will start a new one.";
+    public static final String MESSAGE_FIND_RESULT = "Katheryne: Here are the matching tasks in your list:";
+    public static final String MESSAGE_TASK_NOT_FOUND = "Katheryne: There is no matching task in your list.";
```
I like that your use of constants for common error messages!

hogretup replied:
should these be an enum instead?

hogretup commented on src/main/java/katheryne/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+public class Deadline extends Task{
+    protected LocalDate time;
+
+    public Deadline(String description, String by) {
+        super(description);
+        DateTimeFormatter inputFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+        LocalDate dateObj = LocalDate.parse(by, inputFormatter);
+        this.time = dateObj;
+    }
+
+    public String toString() {
+        DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
+        return String.format("[D]%s(by: %s)", super.toString(), this.time.format(outputFormatter));
+    }
+
+    @Override
+    public String toSaveString() {
+        DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
+        if (isDone) {
+            return String.format("D | %d | %s | %s",1,this.getDescription(),this.time.format(outputFormatter));
+        } else {
+            return String.format("D | %d | %s | %s",0,this.getDescription(),this.time.format(outputFormatter));
```
instead of instantiating a new outputFormatter for each, why not instantiate one for the class?

hogretup commented on src/main/java/katheryne/Katheryne.java:
```diff
@@ -0,0 +1,131 @@
+                } else if (commandWord.equals("mark")) {
+                    System.out.println(c.executeMark(str));
+
+                } else if (commandWord.equals("unmark")) {
+                    System.out.println(c.executeUnmark(str));
+                } else if (commandWord.equals("todo")) {
+                    System.out.println(c.executeAddToDo(str));
+                } else if (commandWord.equals("event")) {
+                    System.out.println(c.executeAddEvent(str));
+                } else if (commandWord.equals("deadline")) {
+                    System.out.println(c.executeAddDeadline(str));
+                } else if (commandWord.equals("bye")) {
+                    isFinish = true;
+                    System.out.println(ui.getDivide());
+                    System.out.println(ui.getBye());
+                    System.out.println(ui.getDivide());
+                } else if (commandWord.equals("delete")) {
+                    System.out.println(c.executeDelete(str));
+                } else if (commandWord.equals("find")) {
+                    System.out.println(c.executeFind(str));
+                } else {
```
very neat!

hogretup commented on src/main/java/katheryne/Parser.java.rej:
```diff
@@ -0,0 +1,16 @@
+ import java.time.format.DateTimeFormatter;
+ import java.time.format.DateTimeParseException;
+ 
++
++/**
++ * Parser class to parse the user input.
++ * Contains methods to parse the user input into commands and descriptions.
++ * Contains methods to parse the user input into date and time.
++ * Contains methods to format the date and time.
++ * Contains methods to check if the date is valid.
++ */
```
good javadocs!

YANYAN-w committed:
Use assertions to check for empty attributes of Task objects

There are places in the Deadline and Event classes where attributes,
such as description or date, might be empty during Task creation.

To ensure consistent validation and improve code readability, we
introduce assertions to verify that essential attributes are not empty.

This change improves code quality by identifying invalid input scenarios
early in the process and reduces redundancy in the Deadline and Event
classes.

YANYAN-w committed:
Refactor command execution methods to  improve readability

Several parts of the command execution logic for the task management
system had duplicated code, particularly in the parsing, validation, and task
creation steps. This commit refactors the code to reduce redundancy and
improve maintainability.

Key changes include:

Extracted the command parsing and validation logic into a reusable
parseAndValidateCommand method. This reduces duplication in methods.

Added a common addTask method to handle task creation for ToDo,
Event, and Deadline tasks. This centralizes task addition logic, making
the code more readable.

Introduced the validateTaskIndex helper method to perform task index
validation in a single place, improving error handling consistency.

Improved the overall structure and readability of the code by reducing
repetition and making method responsibilities more focused.

YANYAN-w committed:
Merge pull request #1 from YANYAN-w/branch-A-Assertions

Use assertions to check for empty attributes of Task objects

YANYAN-w committed:
Merge branch 'master' of https://github.com/YANYAN-w/ip

* 'master' of https://github.com/YANYAN-w/ip:
  Use assertions to check for empty attributes of Task objects

YANYAN-w committed:
Merge pull request #2 from YANYAN-w/branch-A-CodeQuality

Refactor command execution methods to  improve readability

YANYAN-w committed:
Merge branch 'master' of https://github.com/YANYAN-w/ip

* 'master' of https://github.com/YANYAN-w/ip:
  Refactor command execution methods to  improve readability

YANYAN-w committed:
Update exception handling in Command class

YANYAN-w committed:
Add new feature for detecting duplicated tasks

This commit introduces a new feature to detect duplicated tasks in
the task list. The feature checks for tasks with identical descriptions
and relevant attributes based on the task type (e.g., deadline, event)
to ensure that no duplicate tasks are added to the task list.

The hasDuplicates(Task t) method has been added to the TaskList
class, which loops through the existing tasks and compares them with
the new task. The equality comparison has been refined to handle
class-specific attributes, allowing for accurate detection of duplicates
across different task types.

This feature improves the overall task management workflow by preventing redundant tasks from being created.

YANYAN-w committed:
Update for code style check

YANYAN-w committed:
Add image of ui and update the user guide

YANYAN-w committed:
Update README.md for feature explanation

YANYAN-w committed:
Update for code style and add more javadoc commands

YANYAN-w committed:
Update README.md with better visuals
